src/Bicep.Core/Emit/ExpressionConverter.cs

```diff
@@ -142,185 +135,127 @@ public LanguageExpression ConvertExpression(Expression expression)
                     syntax = syntaxExpression.Syntax;
                     break;
 
-                default:
-                    throw new NotImplementedException($"Cannot emit unexpected expression of type {expression.GetType().Name}");
-            }
-
-            switch (syntax)
-            {
-                case UnaryOperationSyntax unary:
+                case UnaryExpression unary:
                     return ConvertUnary(unary);
 
-                case BinaryOperationSyntax binary:
+                case BinaryExpression binary:
                     return ConvertBinary(binary);
 
-                case TernaryOperationSyntax ternary:
+                case TernaryExpression ternary:
                     return CreateFunction(
                         "if",
-                        ConvertExpression(ternary.ConditionExpression),
-                        ConvertExpression(ternary.TrueExpression),
-                        ConvertExpression(ternary.FalseExpression));
+                        ConvertExpression(ternary.Condition),
+                        ConvertExpression(ternary.True),
+                        ConvertExpression(ternary.False));
 
-                case FunctionCallSyntaxBase functionCall:
-                    return ConvertFunction(functionCall);
+                case FunctionCallExpression function:
+                    return CreateFunction(
+                        function.Name,
+                        function.Parameters.Select(p => ConvertExpression(p)));
 
-                case ArrayAccessSyntax arrayAccess:
-                    return ConvertArrayAccess(arrayAccess);
+                case ArrayAccessExpression exp:
+                    return AppendProperties(
+                        ToFunctionExpression(ConvertExpression(exp.Base)),
+                        ConvertExpression(exp.Access));
 
-                case ResourceAccessSyntax resourceAccess:
-                    return ConvertResourceAccess(resourceAccess);
+                case PropertyAccessExpression { Base: ResourceReferenceExpression resource } exp:
+                    return GetConverter(resource.IndexContext).ConvertResourcePropertyAccess(resource, exp);
 
-                case PropertyAccessSyntax propertyAccess:
-                    return ConvertPropertyAccess(propertyAccess);
+                case PropertyAccessExpression { Base: ModuleReferenceExpression module } exp:
+                    return GetConverter(module.IndexContext).ConvertModulePropertyAccess(module.Module, exp);
```

Comment:
Yes, ideally, but I'm trying to avoid modifying existing behavior while doing this refactor. Currently the following results in sub-optimal codegen:

```bicep
resource foo 'Microsoft.Storage/storageAccounts@2022-09-01' existing = {
  name: 'asdf'
}

output asdf string = foo['id']
```

I'll raise a bug - should be fairly straightforward to address once this PR is merged.

Reply:
Raised #9285 for this.

Reply:
Additionally, do we need to maintain docs with the list of transformations that we do from expression to IR? For example when <base>['foo'] is replaced with property access, how does a Bicep dev know that they no longer need to concern themselves with handling constant string indexers on an object.

Reply:
One option could be a comment convention in the IR layer to explicitly call out non-1:1 transforms.